src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,131 +14,184 @@ namespace System.Buffers
+                    goto EndPositionNotReached;
 
-                if (startSegment != endObject)
+                // End segment
+                if (positionObject == endObject)
                 {
-                    ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;
+                    ReadOnlySequenceSegment<T> positionSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(positionObject);
 
-                    if (nextSegment == null && startSegment != endObject)
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    // Bounds check
+                    if (!trusted && positionSegment == _sequenceStart.GetObject() && positionIndex < GetIndex(_sequenceStart))
+                        goto PositionOutOfRange;
 
+                    int length = GetIndex(_sequenceEnd) - positionIndex;
+                    if (!trusted && length < 0)
+                        goto PositionOutOfRange;
+
+                    memory = positionSegment.Memory.Slice(positionIndex, length);
+                    next = default;
+                    return true;
+                }
+
+                // Start or Middle segment
+                {
+                    ReadOnlySequenceSegment<T> positionSegment;
+
+                    if (!trusted) // Bounds check
+                    {
+                        object startObject = _sequenceStart.GetObject();
+                        if (positionObject == startObject) // Start segment
+                        {
+                            positionSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(positionObject);
+                            if (positionIndex < GetIndex(_sequenceStart))
+                                goto PositionOutOfRange;
+                        }
+                        else // Middle Segment
+                        {
+                            positionSegment = positionObject as ReadOnlySequenceSegment<T>;
+                            if (positionSegment == null)
+                                goto EndPositionNotReached;
+
+                            ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
+                            ReadOnlySequenceSegment<T> endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(endObject);
+
+                            if (positionSegment.RunningIndex - startSegment.RunningIndex < 0 || endSegment.RunningIndex - positionSegment.RunningIndex < 0)
+                                goto PositionOutOfRange;
+                        }
+                    }
+                    else
+                    {
+                        positionSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(positionObject);
+                    }
+
+                    ReadOnlySequenceSegment<T> nextSegment = positionSegment.Next;
+                    if (!trusted && nextSegment == null)
+                        goto EndPositionNotReached;
```

Comment:
why not just throw directly?

Reply:
This method is Aggressive inlining so I try to make less calls.

Reply:
you can make exception throwing inline as well or convert it to return exception and throw `CreateExceptionXY()` - not sure if performance for throwing exception is that relevant

Reply:
`if (something1) goto EndPositionNotReached; if (something2) goto EndPositionNotReached;` is just 2 jumps.
`if (something1) ThrowEndPositionNotReached(); if (something2) ThrowEndPositionNotReached();` is 2 jumps and 2 calls.
So first is less.

Reply:
@AlexRadch I do not think it is worth to make source code less readable to save few bytes in the machine code. It makes more sense to send a suggestion to JIT/compiler team to improve such cases

Reply:
Why `goto EndPositionNotReached;` is less readable than `ThrowEndPositionNotReached();` ?
I can change it to `goto ThrowEndPositionNotReached;`